### ISSUE TITLE
[API Proposal]: Analyzer - replace copy loop with span.CopyTo(destination)

### ISSUE THREAD

**Issue by neon-sunset** [dotnet/runtime#99186]
### Background and motivation

Given copy unrolling improvements, using `span.CopyTo(dest)` should outperform manual copy with a loop in almost every situation. It may be desirable to have a performance analyzer rule to flag such use cases and suggest an auto-fixer to address them.

Example pattern spotted in https://github.com/dotnet/runtime/blob/2ab51c6ef3fb24a625c1f4729f9b2e626a0f141b/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L405-L413

### API Proposal

--

### API Usage

```csharp
var src = Enumerable
    .Range(0, 16)
    .Select(num => num.ToString())
    .ToArray()
    .AsSpan()
var dst = new string[16];

// Flagged with a rule "use span.CopyTo for copying elements"
for (var i = 0; i < src.Length; i++)
{
    dst[i] = src[i];
}
```

### Alternative Designs

N/A

### Risks

Should be none (right? :) )

**Comment by EgorBo** [comment]
The only potential problem here that `CopyTo` relaxes atomicity guarantees, so you wrote a code that atomically (per our memory model) writes/copies elements, now you're suggesting to replace it with CopyTo that might produce torn values in a multi-thread environment.

**Comment by neon-sunset** [comment]
This could possibly have heuristics like copying elements below or equal to pointer size will appear with greater or the same degree of atomicity. So no kicking in for structs that could be torn - it will cover most common case like assigning object references, integers, etc.

**Comment by MichalPetryka** [comment]
> This could possibly have heuristics like copying elements below or equal to pointer size will appear with greater or the same degree of atomicity. So no kicking in for structs that could be torn - it will cover most common case like assigning object references, integers, etc. (yes, 32bit systems, but there are tons of bespoke copy loops for integers, chars as linked, and more)

`CopyTo` doesn't provide any atomicity guarantees, on ARM64 it should be atomic for pointer accesses at pointer aligned locations but on X64 it's not guaranteed.

**Comment by neon-sunset** [comment]
>`CopyTo` doesn't provide any atomicity guarantees, on ARM64 it should be atomic for pointer accesses at pointer aligned locations but on X64 it's not guaranteed.

Right, and spans don't provide alignment guarantees for these either (if they did, then there would be atomicity guarantees).

So, the only element types left are `byte` and `object`.

**Comment by huoyaoyuan** [comment]
I'm now curious that: does Span.CopyTo guarantee proper element-wise access for properly aligned spans?

**Comment by EgorBo** [comment]
> I'm now curious that: does Span.CopyTo guarantee proper element-wise access for properly aligned spans?

With the current implementation it does not

**Comment by EgorBo** [comment]
So overall suggestions like these are probably nice to have! (poor man's loop-auto-vectorization) but the user should be aware of risks

**Comment by neon-sunset** [comment]
Thanks for the feedback.

Perhaps this could be a warning for loops where the analyzer sees that atomicity is preserved (bytes and objects) while for other cases it might be a suggestion, further relying on whether the buffer is declared locally and is not observed before the copy is done?

This seems like a fairly widespread mistake to copy the elements manually, and if the users case about atomicity, it is likely they already know which option to go with.
